src/test/java/com/vmware/vip/i18n/LocaleTest.java

```diff
@@ -46,34 +42,72 @@ public void init() {
     public void testPickupLocaleFromList() {
         Locale[] supportedLocales = { Locale.forLanguageTag("de"),
                 Locale.forLanguageTag("es"), Locale.forLanguageTag("fr"),
+                Locale.forLanguageTag("fr-CA"),
                 Locale.forLanguageTag("ja"), Locale.forLanguageTag("ko"),
+                Locale.forLanguageTag("zh"),
                 Locale.forLanguageTag("zh-Hans"),
                 Locale.forLanguageTag("zh-Hant")
 
         };
         Locale[] testLocales = { Locale.forLanguageTag("de"),
                 Locale.forLanguageTag("es"), Locale.forLanguageTag("fr"),
+                Locale.forLanguageTag("fr-CA"), Locale.forLanguageTag("fr-FR"),
                 Locale.forLanguageTag("ja"), Locale.forLanguageTag("ko"),
-                Locale.forLanguageTag("zh"), Locale.forLanguageTag("zh-CN"),
+                Locale.forLanguageTag("zh"),
+                Locale.forLanguageTag("zh-CN"),
                 Locale.forLanguageTag("zh-TW"),
                 Locale.forLanguageTag("zh-HANS-CN"),
                 Locale.forLanguageTag("zh-HANT-TW"),
                 Locale.forLanguageTag("zh-HANS"),
                 Locale.forLanguageTag("zh-HANT") };
 
-        String[] expectedLocales = { "de", "es", "fr", "ja", "ko", "zh",
+        String[] expectedLocales = { "de", "es", "fr", "fr-CA", "fr", "ja", "ko", "zh",
                 "zh-Hans", "zh-Hant", "zh-Hans", "zh-Hant", "zh-Hans", "zh-Hant" };
 
         for (int i = 0; i < testLocales.length; i++) {
             String matchedLanguageTag = LocaleUtility.pickupLocaleFromList(
-                    Arrays.asList(supportedLocales), testLocales[i])
+                    new HashSet<>(Arrays.asList(supportedLocales)), testLocales[i])
                     .toLanguageTag();
 
             logger.debug(matchedLanguageTag + "-----" + expectedLocales[i]);
             Assert.assertEquals(expectedLocales[i], matchedLanguageTag);
         }
     }
 
+    @Test
+    public void testPickupLocaleFromListNotFound() {
+        Locale[] supportedLocales = { Locale.forLanguageTag("de"),
+                Locale.forLanguageTag("es"), Locale.forLanguageTag("fr"),
+                Locale.forLanguageTag("fr-CA"),
+                Locale.forLanguageTag("ja"), Locale.forLanguageTag("ko"),
+                Locale.forLanguageTag("zh-Hans"),
+                Locale.forLanguageTag("zh-Hant")
+
+        };
+        Assert.assertNull(LocaleUtility.pickupLocaleFromList(new HashSet<>(Arrays.asList(supportedLocales)), Locale.forLanguageTag("fil")));
+    }
+
+    /**
+     *  For any Chinese locale (zh-*) that is not supported,
+     *  return null so that fallback locale will be used even if "zh" is supported.
+     *  For any non-Chinese locale, return the best match (e.g. 'de' if 'de-DE' is not supported).
+     */
+    @Test
+    public void testPickupLocaleFromList_special_case_zh_HK() {
```

Comment:
Add one more test case: zh is supported, zh-Hans and zh-Hant aren't supported.  Request zh-Hant will return en-US.